sherpa/astro/flux.py

```diff
@@ -615,88 +614,55 @@ def calc_sample_flux(id, lo, hi, session, fit, data, samples, modelcomponent,
+    assert nthawed == npars, (nthawed, npars)
 
-    orig_source = session.get_source(id)
+    # Identify any row where a parameter lies outside the min/max range
+    softmins = fit.model.thawedparmins
+    softmaxs = fit.model.thawedparmaxes
 
-    logger = logging.getLogger("sherpa")
-    orig_log_level = logger.level
+    # We have to use columns 1 to n-1 of samples
+    valid = numpy.ones(samples.shape[0], dtype=numpy.bool)
+    for col, pmin, pmax in zip(samples.T[1:-1], softmins, softmaxs):
+        valid &= (col >= pmin) & (col <= pmax)
 
-    # only change the log level if it is less than error
-    #
-    if orig_log_level < logging.ERROR:
-        logger.setLevel(logging.ERROR)
+    nrows = samples.shape[0]
+    oflx = samples[:, 0]       # observed/absorbed flux
+    iflx = numpy.zeros(nrows)  # intrinsic/unabsorbed flux
 
+    mystat = numpy.zeros((nrows, 1), dtype=samples.dtype)
     try:
+        for nn in range(nrows):
+            # Need to extract the subset that contains the parameters
+            fit.model.thawedpars = samples[nn, 1:-1]
+            if valid[nn]:
+                iflx[nn] = calc_energy_flux(data, modelcomponent, lo=lo, hi=hi)
 
-        softmins = fit.model.thawedparmins
-        softmaxs = fit.model.thawedparmaxes
-        mysim = simulated_pars_within_ranges(samples, softmins, softmaxs)
-
-        size = len(mysim[:, 0])
-        oflx = numpy.zeros(size)  # observed/absorbed flux
-        iflx = numpy.zeros(size)  # intrinsic/unabsorbed flux
-        thawedpars = [par for par in fit.model.pars if not par.frozen]
-
-        logger.setLevel(logging.ERROR)
-
-        mystat = []
-        for nn in range(size):
-            session.set_source(id, orig_source)
-            oflx[nn] = mysim[nn, 0]
-            for ii in range(len(thawedpars)):
-                val = mysim[nn, ii + 1]
-                session.set_par(thawedpars[ii].fullname, val)
-            session.set_source(id, modelcomponent)
-            iflx[nn] = session.calc_energy_flux(lo=lo, hi=hi, id=id)
-            #####################################
-            session.set_full_model(id, orig_model)
-            mystat.append(session.calc_stat(id))
-            #####################################
-
-        logger.setLevel(orig_log_level)
+            mystat[nn, 0] = fit.calc_stat()
 
     finally:
-
-        # Why do we set both full_model and source here?
-        #
-        logger.setLevel(logging.ERROR)
-        session.set_full_model(id, orig_model)
-        fit.model.thawedpars = orig_model_vals
-        session.set_source(id, orig_source)
-
-        logger.setLevel(orig_log_level)
+        fit.model.thawedpars = thawedpars
 
     hwidth = confidence / 2
     result = []
-    for x in [oflx, iflx]:
-        result.append(numpy.percentile(x, [50, 50 + hwidth, 50 - hwidth]))
+    for flx in [oflx, iflx]:
+        result.append(numpy.percentile(flx[valid],
+                                       [50, 50 + hwidth, 50 - hwidth]))
```

Comment:
So, `oflx` is just the same as `iflx`. Why have two different variables in that case?

Reply:
They are only the same when modelcomponent is None. When modelcomponent is set iflx is set to the flux of modelcomponent. We could make this code copy over the oflx case when modelcomponent is not set, but I thought it would probably just be harder to understand, but I admit to not actually giving it much thought.

Reply:
That's OK, I just wanted to understand better.